docs/design/federation-phase-1.md

```diff
@@ -0,0 +1,262 @@
++  Cluster Size Enhancement. Currently K8S cluster can only support a limited size. While the community is actively improving it, it can be expected that cluster size will be a problem if K8S is used for large workloads or public PaaS infrastructure. While we can separate different tenants to different clusters, it would be good to have a unified view. 
+Here are the functionality requirements derived from above use cases:
+
++  Clients can register and deregister clusters.
++  Workloads should be spread to different clusters according to the workload distribution policy.
++  Pods running on different clusters are able to discover each other and communicate with each other.
++  Traffic to these pods should be load balanced among clusters. 
++  The control plane needs to know when a cluster is down, and migrate the workloads to other clusters.
++  Clients have a unified view and a central control point for above activities.
+
+## SCOPE
+
+It’s difficult to have a perfect design with one click that implements all the above requirements. Therefore we will go with an iterative approach to design and build the system. This document describes the phase one of the whole work.   
+In phase one we will cover only the following objectives:
+
++  Define the basic building blocks and API objects of control plane
++  Implement a basic end-to-end workflow
+   +  Clients register federated clusters
+   +  Clients submit a workload
+   +  The workload is distributed to different clusters 
+   +  Service discovery
+   +  Load balancing
+The following parts are NOT covered in phase one:
+
++  Authentication and authorization
++  Deployment units other than replication controller
++  Complex distribution policy of workloads
++  Service affinity and migration 
+
+## ARCHITECTURE
+
+The overall architecture of a control plane is shown as following:
+
+![image](img.jpg)
+
+Some design principles we are following in this architecture:
+
+1. Keep the underlying K8S clusters independent. They should have no knowledge of control plane or of each other.
+1. Keep the Ubernetes API interface compatible with K8S API as much as possible.
+1. Re-use concepts from K8S as much as possible. This reduces customers’ learning curve and is good for adoption.
+Below is a brief description of each module contained in above diagram. 
+
+## API Server
+
+The API Server in Ubernetes control plane works just like the API Server in K8S. It talks to the distributed storage to persist, retrieve and watch API objects.  
+We still use ETCD as the distributed storage so customers don’t need to learn and manage a different storage system.
+
+## Scheduler
+
+The scheduler watches pending replication controllers and performs the global scheduling work.  
+For each pending replication controller, it calls policy engine to decide how to spit workloads among clusters. It creates sub-RCs and post them back to ETCD storage.
```

Comment:
@lavalamp Point taken that this is perhaps too hand-wavey.  I will try to improve upon it.  The important part is that there needs to be a distinction between the policy that the federation administrator needs to be able to configure (e.g. "favor cluster provider X over cluster provider Y, because it's cheaper"), and the properties of the Pod/Replication Controller specified by the application administrator (e.g. "this pod runs correctly on cluster providers X and Y, but not Z, where it just crashes").